types/trigger.d.ts

```diff
@@ -43,34 +46,118 @@ export type Output = {
 
 export type TriggerAutoConfig = {
   Output?: Output;
   Duration?: number;
   BeforeSeconds?: number;
   OutputStrings?: { [outputKey: string]: Lang };
 }
+
+export type MatchesAny = { [s in T]?: string } | undefined;
+
+// Note: functions like run or preRun need to be defined as void-only as (confusingly)
+// it is not possible to assign `(d: Data) => boolean` to a void | undefined, only to void.
+type OptionalUnlessVoid<T> = T extends void ? void : T | undefined;
+export type TriggerField<Data, Return> =
+    TriggerFunc<Data, MatchesAny, OptionalUnlessVoid<Return>> | OptionalUnlessVoid<Return>;
+
+// TODO: I am not sure that it is possible to type triggers such that we have some gigantic union
+// of Trigger<TetherParams> | Trigger<AbilityParams> | Trigger<AddedCombatantParams> | etc etc
+// and so the idea is that individual files would use a narrower type for Matches if they wanted to
+// and that's something we (maybe?) could enforce via tests rather than via TypeScript.  This
+// approach would not preclude some TypeScript gigabrain solution that can figure out how to do
+// the same thing at the TypeScript level.
+//
+// TODO: I think there are two types of TriggerSets.  There is the "cactbot + user defined triggers"
+// and also the "cactbot built-in only" triggers, which can be more strict.  For example,
+// * is `id` required, and `regex` for timeline triggers? (yes for built-in, no for user triggers)
+// * should we handle nonsense like `condition` returning non-booleans?
+// We could have one type that built-in triggers define themselves as more strictly and then a
+// second type that popup-text.js uses and needs to do more narrowing for.
+export type BaseTrigger<Data> = {
+  id?: string;
+  disabled?: boolean;
+  condition?: TriggerField<Data, boolean>;
+  preRun?: TriggerField<Data, void>;
+  delaySeconds?: TriggerField<Data, number>;
+  durationSeconds?: TriggerField<Data, number>;
+  suppressSeconds?: TriggerField<Data, number>;
+  promise?: TriggerField<Promise<Data, void>>;
+  sound?: TriggerField<Data, string>;
+  soundVolume?: TriggerField<Data, number>;
+  response?: ResponseField<Data>;
+  alarmText?: TriggerField<Data, TriggerOutput>;
+  alertText?: TriggerField<Data, TriggerOutput>;
+  infoText?: TriggerField<Data, TriggerOutput>;
+  tts?: TriggerField<Data, TriggerOutput>;
+  run?: TriggerField<Data, void>;
+  outputStrings?: {
+    [key: string]: LocaleText;
+  };
+}
+
+export type NetRegexTrigger<Data> = BaseTrigger<Data> & {
+  netRegex: RegExp;
+  netRegexDe?: RegExp;
+  netRegexFr?: RegExp;
+  netRegexJa?: RegExp;
+  netRegexCn?: RegExp;
+  netRegexKo?: RegExp;
+}
+
+export type RegexTrigger<Data> = BaseTrigger<Data> & {
+  regex: RegExp;
+  regexDe?: RegExp;
+  regexFr?: RegExp;
+  regexJa?: RegExp;
+  regexCn?: RegExp;
+  regexKo?: RegExp;
+}
+
+export type TimelineTrigger<Data> = BaseTrigger<Data> & {
+  regex?: RegExp;
```

Comment:
how will happen if `TimelineTrigger` doesn't have a `regex` property? 
will it match all timeline record?

Reply:
Then the current trigger would be ignored.
I think this should be a required field?

Reply:
Ok, I separated out the types into "loose" types (aka things that might be from users, where we don't want to assume that fields are set, meant to be used in like timeline/popup-text for processing triggers) and into "strict" types (where more things are required, deprecated fields are removed, and meant to be used in built-in triggers).